lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -113,40 +113,74 @@ def _put(self, params: ApiParams, _) -> Response:
         }
         """
 
-        try:
-            key = params.uri["key"]
-            try:
-                value = params.query["value"]
-            except KeyError:
-                try:
-                    value = params.body["value"]
-                except KeyError:
+        # First, we check for the ways to set an individual config setting by
+        # naming the config key on the URI and specifying a value either with
+        # the "value" either as a query parameter or in the JSON request body.
+        #
+        # We'll complain about JSON request body parameters that are "shadowed"
+        # by the URI or query parameter and might represent client confusion.
+        keydup = set()
+        key = params.uri.get("key")
+        if key:
+            # If we have a key in the URL, the we need a "value" for it, which
+            # we can take either from a query parameter or from the JSON
+            # request payload.
+            if params.body and list(params.body.keys()) != ["value"]:
+                keydup.update(params.body.keys())
+            value = params.query.get("value")
+            if value:
+                if params.body:
+                    keydup.update(params.body.keys())
+            else:
+                value = params.body.get("value")
+                if not value:
                     raise APIAbort(
                         HTTPStatus.BAD_REQUEST,
-                        "Incorrect value specification for key {key!r}",
+                        f"No value found for key system configuration key {key!r}",
                     )
+
+            if len(keydup) > 0:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Redundant parameters specified in the JSON request body: {sorted(keydup)!r}",
+                )
+
             try:
                 ServerConfig.set(key=key, value=value)
                 return jsonify({key: value})
             except ServerConfigBadValue as e:
                 raise APIAbort(HTTPStatus.BAD_REQUEST, str(e))
             except ServerConfigError as e:
                 raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
-        except KeyError:
-            pass
-
-        try:
-            config = params.body["config"]
-        except KeyError:
-            raise MissingParameters(["config"])
 
-        failures = []
-        for k, v in config.items():
-            try:
-                ServerConfig.set(key=k, value=v)
-            except ServerConfigError as e:
-                self.logger.warning("Unable to update key {} = {!r}: {}", k, v, str(e))
-                failures.append(k)
-        if failures:
-            raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR)
-        return jsonify(ServerConfig.get_all())
+        else:
```

Comment:
I don't think we need this `else` statement -- if we enter the "then" block, then we either `return` or `raise`.

But, more significantly, this function has gotten really large -- I suggest putting the "then" block in its own subroutine and putting the `else` block in its own subroutine.  Yeah, that will make this function nearly trivial, but as it stands, unit testing this stuff is going to be a bunch of work.

Reply:
I did it this way because you had previously been confused due to missing that the first block only returned or raised. I figured the explicit `else`, while technically redundant, made this more obvious. I'm not entirely sure that breaking into a subroutine makes it easier to read, either.

Reply:
> I did it this way because you had previously been confused due to missing that the first block only returned or raised. I figured the explicit else, while technically redundant, made this more obvious.

That's fair.  (And, thanks!  🙂)

> I'm not entirely sure that breaking into a subroutine makes it easier to read, either.

Well, it makes it clear that the first part won't "fall through" into the second part, because they are in different functions.  And, it reduces all of the indentation by at least one level (which helps, too).  It also reduces the scope that the reader has to consider, and hopefully allows it "all" to fit in one window.  (And, the knock on of that is that it makes the subsequent unit testing simpler, by removing the initial conditional from the execution path.)